WolvenKit/Views/Editors/AutomaticNodifyEditor.cs

```diff
@@ -0,0 +1,230 @@
+                {
+                    Focus();
+                    CaptureMouse();
+
+                    InitialMousePosition = e.GetPosition(this);
+                }
+            }
+            else
+            {
+                base.OnMouseDown(e);
+            }
+        }
+
+        protected override void OnMouseUp(MouseButtonEventArgs e)
+        {
+            if (e.ChangedButton == MouseButton.Middle)
+            {
+                // If right clicking without panning or selecting allow context menu
+                if (IsMouseCaptured && !IsPanning && !IsSelecting)
+                {
+                    Focus();
+                    ReleaseMouseCapture();
+                }
+                // If is selecting but not panning, end selection and show context menu
+                else if (IsSelecting && !IsPanning)
+                {
+                    Selection.End();
+                }
+                else if (IsPanning)
+                {
+                    IsPanning = false;
+
+                    // Allow selecting and panning at the same time and disable context menu
+                    if (IsSelecting)
+                    {
+                        e.Handled = true;
+                    }
+                    // If is panning but not selecting, release mouse capture and show context menu if necessary
+                    else
+                    {
+                        // Handle right click if is panning and moved the mouse more than threshold so context menus don't open
+                        if ((CurrentMousePosition - InitialMousePosition).LengthSquared > HandleRightClickAfterPanningThreshold * HandleRightClickAfterPanningThreshold)
+                        {
+                            e.Handled = true;
+                        }
+
+                        if (IsMouseCaptured)
+                        {
+                            ReleaseMouseCapture();
+                        }
+                    }
+                }
+            }
+            else
+            {
+                base.OnMouseUp(e);
+            }
+        }
+
+        /// <summary>
+        /// Uses <see cref="Microsoft.Msagl.Core"/> to automatically lay out nodes, ignoring socket positions.
+        /// </summary>
+        /// <returns>If the lay out was successful</returns>
+        public bool LayoutNodes()
+        {
+            UpdateLayout();
+            var graph = new GeometryGraph();
+            var nodes = new Dictionary<int, Microsoft.Msagl.Core.Layout.Node>();
+            var socketNodeLookup = new Dictionary<int, int>();
+            foreach (var item in ItemsHost.Children)
+            {
+                if (item is ItemContainer ic)
+                {
+                    if (ic.DataContext is INode nvm)
+                    {
+                        // without UpdateLayout(), these were 0, so backups were needed
+                        var width = ic.ActualWidth != 0 ? ic.ActualWidth : 300;
```

Comment:
You can bind to the `ActualSize` of the `ItemContainer` in v2.0.0 and move the layout logic into the view model.

Reply:
Here's an example of properties that you can bind to the editor if you need more control over it. And there are also some useful methods exposed. https://github.com/miroiu/nodify/blob/feature/new-blueprint/Nodifier/Graph/Graph.Editor.cs